### ISSUE TITLE
Displacement field acquisition issue

### ISSUE THREAD

**Issue by WenhongLan** [SimpleITK/SimpleITK#1425]
Hello

I am the beginner of Simple ITK and I am working on a small project for 2D image registration.

I have an issue that I tried to acquired the DVF from the BSpline transformation and the DVF only has the vertical direction and the horizontal value is almost zero. I compared the result with that by other solution and the DVF is incorrect. I have several series data. But only one of them is not working for the DVF. I checked the image and its metadata carefully. The problem seems not from the data itself.

The image of incorrect DVF is below:
![Picture 1](https://user-images.githubusercontent.com/81567526/126360550-c7a12da3-403a-475d-93f0-e8ce0c1a7ed5.png)

This is a reasonable image with DVF I get from other data.
![Picture 2](https://user-images.githubusercontent.com/81567526/126361228-378ba6c7-4deb-4cfe-962b-6cf784873371.png)

I show the all code below:
Registration key information: BSpline Registration and Set metric as mean squares.
 --------
# the code is based on the example of SimpleITK, whose url is:
# https://simpleitk.readthedocs.io/en/latest/link_ImageRegistrationMethodBSpline3_docs.html

import SimpleITK as sitk
import os
import sys
import numpy as np
import cv2
from cv2 import cv2
from matplotlib import pyplot as plt
from SimpleITK import Transform

def command_iteration(method, bspline_transform):
    if method.GetOptimizerIteration() == 0:
        # The BSpline is resized before the first optimizer
        # iteration is completed per level. Print the transform object
        # to show the adapted BSpline transform.
        print(bspline_transform)

    print(f"{method.GetOptimizerIteration():3} = {method.GetMetricValue():10.5f}")

def command_multi_iteration(method):
    # The sitkMultiResolutionIterationEvent occurs before the
    # resolution of the transform. This event is used here to print
    # the status of the optimizer from the previous registration level.
    if R.GetCurrentLevel() > 0:
        print(f"Optimizer stop condition: {R.GetOptimizerStopConditionDescription()}")
        print(f" Iteration: {R.GetOptimizerIteration()}")
        print(f" Metric value: {R.GetMetricValue()}")

    print("--------- Resolution Changing ---------")

#---------------------------------
# generate the pathway to the data
#---------------------------------
index_volter = '4' # index of the volunteer.
index_case = '3' # index of the scanning.

temp_volter = 'volunteer ' + index_volter
temp_case = 'case ' + index_case
path1=os.path.abspath('.')
#temp_path= os.path.join(path1,'data',temp_volter,'pre-processing',temp_case,'MR images','navi_non_axis_enhanced','')
temp_path= os.path.join(path1,'data',temp_volter,'pre-processing',temp_case,'MR images','navi_rename_IMA','')

file_list = os.listdir(temp_path) # this is the list of the file name.
file_list.sort(key=lambda x:int(x.split('.')[0])) # to sort the file name in positive order

# set the fixed images
fixed_ima_list = os.path.join(temp_path,file_list[0])
fixed_before_mask = sitk.ReadImage(fixed_ima_list,sitk.sitkFloat32) # variable of the fixed image.

#---------------------------------------------
# filter for converting 3D volume to 2D images (fixed images)
#---------------------------------------------
fixed_2D = fixed_before_mask

origin_size = fixed_2D.GetSize()
new_size=[origin_size[0],origin_size[1],0]
filter_2D = sitk.ExtractImageFilter()
filter_2D.SetSize(new_size)
fixed_2D = filter_2D.Execute(fixed_2D)
Median_filter = sitk.MedianImageFilter()
Median_filter.SetRadius(2)
fixed_2D=Median_filter.Execute(fixed_2D)
# comment: because the depth is only 1. Thus, it doesn't matter for the depth

#-----------------------------------------
# setup a mask for ROI region of interest.
#-----------------------------------------
mask_shape = sitk.GetArrayFromImage(fixed_2D)
mask_zero_shape = np.zeros(mask_shape.shape)
mask_zero_shape[30:180,35:125] = 1

mask_bg = sitk.sitkInt8
mask_bg = sitk.GetImageFromArray(mask_zero_shape, isVector=False) # the mask is covered from row 20 to row 140 of the image.
# The following 2 raws are to guarantee the mask's origin, spacing and dimensions are the same as the fixed and 
# moving images (2D images)
mask_bg.SetOrigin(fixed_2D.GetOrigin())
mask_bg.SetSpacing(fixed_2D.GetSpacing()) 
mask_bg.SetDirection(fixed_2D.GetDirection()) 
# apply the mask
fixed = sitk.Mask(fixed_2D, sitk.Cast(mask_bg,sitk.sitkInt8))

k = 0
itrable_k = range(0, len(file_list))
for k in itrable_k:
    moving_ima_list = os.path.join(temp_path,file_list[k])
    moving_before_mask = sitk.ReadImage(moving_ima_list,sitk.sitkFloat32) # moving images.
    moving_2D = moving_before_mask
    moving_2D = Median_filter.Execute(moving_2D)
    moving_2D = filter_2D.Execute(moving_2D)

    moving = sitk.Mask(moving_2D, sitk.Cast(mask_bg,sitk.sitkInt8))

    transformDomainMeshSize = [2] * fixed.GetDimension() # get the size of fixed images.

    #-----------------------
    # BSpline transformation
    #-----------------------
    # use the BSpline Method for image registration.
    # it returns a transformed matrix (mapping) between the moving and fixed images.
    tx = sitk.BSplineTransformInitializer(fixed,transformDomainMeshSize) 

    print(f"Initial Number of Parameters: {tx.GetNumberOfParameters()}")

    transform_output_temp = os.path.join(path1,'data',temp_volter,'pre-processing',temp_case,'MR images','backup',\
        'SimpleITK','with_Mask','navi','transformation')
    if not os.path.exists(transform_output_temp): # create the saving path for transformation
        os.mkdir(transform_output_temp)
    transform_name = str(k) + '.tfm'
    transform_output_path = os.path.join(transform_output_temp, transform_name)
    #----------------------------------------------
    # configurate the image registration parameters
    #----------------------------------------------
    R = sitk.ImageRegistrationMethod()
    R.SetMetricAsMeanSquares()  # important

    R.SetOptimizerAsGradientDescentLineSearch(5.0,
                                              100,
                                              convergenceMinimumValue=1e-4,
                                              convergenceWindowSize=5)

    R.SetInterpolator(sitk.sitkLinear) # important

    R.SetInitialTransformAsBSpline(tx,
                                   inPlace=True,
                                   scaleFactors=[1, 2, 5])
    R.SetShrinkFactorsPerLevel([16, 9, 4, 2])
    R.SetSmoothingSigmasPerLevel([16, 9, 4, 2])

    R.AddCommand(sitk.sitkIterationEvent, lambda: command_iteration(R, tx))
    R.AddCommand(sitk.sitkMultiResolutionIterationEvent,
                 lambda: command_multi_iteration(R))

    outTx = R.Execute(fixed, moving) # 

    print("-------")
    print(tx)
    print(outTx)
    print(f"Optimizer stop condition: {R.GetOptimizerStopConditionDescription()}")
    print(f" Iteration: {R.GetOptimizerIteration()}")
    print(f" Metric value: {R.GetMetricValue()}")

    sitk.WriteTransform(outTx, transform_output_path) # write down the transformation to the specific path

    #-------------------
    # displacement field
    #-------------------
    displ = sitk.TransformToDisplacementField(outTx, 
                                  sitk.sitkVectorFloat64,
                                  fixed_2D.GetSize(),
                                  fixed_2D.GetOrigin(),
                                  fixed_2D.GetSpacing(),
                                  fixed_2D.GetDirection())
    

    save_path_displ = os.path.join(path1,'data',temp_volter,'pre-processing',temp_case,'MR images','backup',\
        'SimpleITK','with_Mask','navi','displ')
    if not os.path.exists(save_path_displ):
            os.mkdir(save_path_displ)
    name_displ = str(k) + '.nii'
    save_path_displ_full = os.path.join(save_path_displ,name_displ)
    sitk.WriteImage(displ,save_path_displ_full)

    #-----------------------
    #plot and save the image
    #-----------------------
    if ("SITK_NOSHOW" not in os.environ):
        resampler = sitk.ResampleImageFilter()
        resampler.SetReferenceImage(fixed)
        resampler.SetInterpolator(sitk.sitkLinear)
        resampler.SetDefaultPixelValue(100)
        resampler.SetTransform(outTx)

        out = resampler.Execute(moving)

        # Adapt the histogram equalization
        sitk.AdaptiveHistogramEqualization(sitk.RescaleIntensity(fixed))
        simg1 = sitk.Cast(sitk.RescaleIntensity(fixed), sitk.sitkUInt8)
        simg2 = sitk.Cast(sitk.RescaleIntensity(out), sitk.sitkUInt8)
        cimg = sitk.Compose(simg1, simg2, simg1 // 2. + simg2 // 2.)

        #save path
        temp_path_save = os.path.join(path1, 'data',temp_volter,'pre-processing',temp_case,'MR images','backup',\
            'SimpleITK','with_Mask','navi','registration navi')
        if not os.path.exists(temp_path_save): # create the saving path for registration images
            os.mkdir(temp_path_save)
        navi_save_name = str(k) + '.png'
        temp_path_save2 = os.path.join(temp_path_save, navi_save_name)
        sitk.WriteImage(cimg, temp_path_save2)

    k = k + 1

**Comment by WenhongLan** [comment]
The data is my personal data. So it is no problem of it for show in public.

**Comment by zivy** [comment]
Hello @WenhongLan,

In the future, please provide a minimal working example of your code, otherwise it becomes very hard to follow along with such a lengthy program. 

Having said that, after going through the code, it appears that you are using the  `GradientDescentLineSearch` in combination with the `BSplineTransform`. This combination is not ideal. The optimizers that better deal with the BSplineTransform are `LBFGS` and `LBFGS2`, as they work well with a high number of parameters. The `LBFGS2` is better suited for changes in the control grid resolution associated with using a pyramid, your case. For a detailed example, please see [this jupyter notebook](https://github.com/InsightSoftwareConsortium/SimpleITK-Notebooks/blob/master/Python/65_Registration_FFD.ipynb).

**Comment by WenhongLan** [comment]
Thank you very much. I will follow the minimal example of my code next time.

And Thanks for your suggestions, I think I found the issue of it, based on your suggestions.

Best regards
Wenhong
